homeassistant/components/samsungtv/media_player.py

```diff
@@ -348,10 +315,10 @@ def media_previous_track(self):
             await asyncio.sleep(KEY_PRESS_TIMEOUT, self.hass.loop)
         await self.hass.async_add_job(self.send_key, "KEY_ENTER")
```

Comment:
If you want to do more cleanup you can change to use `hass.async_add_executor_job` for all these send key calls.

Reply:
Wouldn't be to change all the methods to `async` the better way?

Reply:
You mean make `send_key` a coroutine?

Reply:
We're doing sync work in `send_key` so either we have to schedule that method on the executor or schedule all calls that are sync inside `send_key` on the executor. It's best to have as few context switches as possible.

Reply:
Yes.

Reply:
If the device library is sync we prefer to use the home assistant sync api if possible.

In our case the device library is sync but some entity methods need to use `asyncio.sleep` or call the script.async_run method so have to use the async api for that. For all other methods we should stick with the sync api.

Eg select_source looks like it can be changed to sync api.

Reply:
It's totally fine to save this clean up for another PR. This PR is already big.

Reply:
Ok, this is all a bit confusing. Is there some kind of documentation describing the differences of this functions?
- `add_job` - sync call of `async_add_job`?
- `async_add_job`
- `async_create_task`
- `async_add_executor_job`
- `async_run_job`

Reply:
1. yes, but without possibility to wait for the task.
2. Can handle both async and sync jobs. Legacy for general use. Still needed in some core places where we can't predict the type of job.
3. Schedule coroutines. Returns the task object.
4. Schedule sync jobs on executor thread pool. Returned task can be awaited.
5. Legacy for general use. Just used in some core places.

Reply:
A best practice description here would be helpful: https://developers.home-assistant.io/docs/en/asyncio_categorizing_functions.html#the-callback-function
Should I file a feature request for it?

Reply:
A PR is welcome, not a feature request.

Reply:
I need this as a reminder.